compiler/rustc_trait_selection/src/traits/project.rs

```diff
@@ -2337,9 +2338,24 @@ fn confirm_fn_pointer_candidate<'cx, 'tcx>(
         sig,
     );
 
-    confirm_callable_candidate(selcx, obligation, sig, util::TupleArgumentsFlag::Yes)
-        .with_addl_obligations(nested)
-        .with_addl_obligations(obligations)
+    let host_effect_param = match *fn_type.kind() {
+        ty::FnDef(def_id, args) => tcx
+            .generics_of(def_id)
+            .host_effect_index
+            .map_or(tcx.consts.true_, |idx| args[idx].expect_const()),
```

Comment:
```suggestion
            .map_or(tcx.consts.true_, |idx| args.const_at(idx)),
```

should give a better message when it isn't a const.